### PR TITLE
See and Edit Selected Links in Callout Dialog 

### PR DESCRIPTION
**Pull request by VasuBhog**
This PR addresses #14724. 
It will show the user the currently selected link within the Callout dialog so that the user can see the current link and edit it.

Link-in-CallOut

### CODE REVIEW

File: src/sql/workbench/contrib/notebook/browser/calloutDialog/linkCalloutDialog.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/calloutDialog/linkCalloutDialog.ts:48]: The link isn't really a label - I'd suggest `_defaultLinkUrl` to match the name of the input

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:299]: Same here - `defaultLinkUrl`

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:322]: `getCurrentLinkUrl`

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: Does getSelection also handle just having the cursor in the middle of the link (not selecting any text)? I would expect that even if I don't have anything selected it would still edit that link. 

The comment for the function indicates it does but I just want to make sure that's how it behaves.

And then if it does do that did you test what happens when that happens? Will it update the link as expected or does it try to insert a new link at that location?

Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: It looked like in your example your path started with a / - e.g. `/C:/my/path/notebook.ipynb`. Was this expected?

**Review by Charles-Gagnon — COMMENTED** [review]
Waiting on confirmation of a few things before I sign off.

Also some tests would be great here.

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: I believe that the intended scenario is that the user would select the entire link, as the link does get the link address wherever selected on the link. However, the display text does not include the title, which would require additional work.

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: Yes this is the absolute path, similar to that of what I copied into the link originally.

Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: I assume this is a windows machine - in which case /C:/my/path wouldn't be a valid path I'd expect. You're saying you entered in the link as `/C:/my/path`? Does that link actually work? Regardless - if you did could you show a gif of updating a link without the / prefix?

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (MaddyDev) [inline src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:326]: >document.getSelection() [](start = 10, length = 23)

would this always return a valid anchorNode which will have a parentNode or shoukd be checking for nulls?

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: Not sure I understand - could you get a gif of the behavior for this? I don't think many people will know necessarily to select the entire text (and what happens if they select more/less) so we should probably spend some time now making sure we at least don't break stuff when that happens.

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: Well, I went based on the issue that Julie shows in the [gif](https://github.com/microsoft/azuredatastudio/issues/14724). I think that is the clear scenario that we should address. Addressing the selection of the link regardless of where the cursor is and replacing the entire link based on any location of the link seems to be another piece of work entirely.

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:326]: If the link is selected as shown in the gif in the issue, then there should always be an anchorNode with a parentNode as far as I tested.

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: I agree, I'm mostly saying make sure you test these other scenarios to verify that they aren't broken and then make sure you communicate expectations to the Livesite team so they know what exactly is supported.

**Review by MaddyDev — APPROVED** [review]

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: It is a valid path, as it is the absolute path that can be used for links to a notebook. They get converted to a relative path in markdown preview, so the link does open correctly and is formatted as such in Markdown.
`[WYSIWYG File Test](.\WYSIWYG-test.ipynb)

`fileLinkWYSIWYG

Comment (lucyzhang929) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: Julie also mentioned this in the issue:
"2. Users can place a cursor in the middle of the_Display Text_, and click on the Link button, then show the current values for both Text to display and Address (currently only shows Text to Display value). Changing the Text to display and Address values, will update the Display Text and the URL."
So it would probably be good to check with her or create a new issue to track the second part.

Comment (MsSQLGirl) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: "2. Users can place a cursor in the middle of the_Display Text_, and click on the Link button, then show the current values for both Text to display and Address (currently only shows Text to Display value). Changing the Text to display and Address values, will update the Display Text and the URL."

I would consider this as the next step (incremental). It is possible to ship without it for now.

Comment (MsSQLGirl) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: @VasuBhog - if a link has been resolved to relative link, can you please confirm that updating the link (with the copy path / full path) would still work OK? i.e. the second relative link translation will still happen.

**Comment by Charles-Gagnon** [comment]
@VasuBhog You have some compile errors :

`[08:36:55] Error: /home/runner/work/azuredatastudio/azuredatastudio/src/sql/workbench/contrib/notebook/test/calloutDialog/linkCalloutDialog.test.ts(15,1): 'DOM' is declared but its value is never read.
`

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:324]: @MsSQLGirl, yes if a user enters a new path via the (Copy Path / Full Path) it will be resolved to relative and work as expected.

**Comment by barbaravaldez** [comment]
@VasuBhog - can you check that it works with links to a section for instance `#hello `?

**Comment by VasuBhog** [comment]
> @VasuBhog - can you check that it works with links to a section for instance `#hello `?

Yes links to a header within the notebook still work properly, even after selecting and replacing.

**Review by Charles-Gagnon — APPROVED** [review]

**Pull request merged**
